File: shared/streak_analysis.py
"""
SeasonAlpha — Streak-Analyse (Wiederverwendbar)
=================================================
Zeigt aktuelle Gewinn-/Verlust-Serien als HTML-Tabelle.

Verwendung:
    from shared.streak_analysis import render_streak_table

    # Variante 1: Fertige Gruppen (z.B. Wochentage, Mondphasen)
    render_streak_table(groups, title="Wochentag")

    # Variante 2: Aus DataFrame berechnen
    groups = compute_streaks_from_df(df, group_col="weekday", label_map=LABELS)
    render_streak_table(groups)

Einsatzgebiete:
    - Wochentage (Mo-Fr / Mo-So)
    - Monatswechsel (Jan→Feb ... Dez→Jan)
    - Mondphasen (Vollmond, Neumond)
    - OPEX (monatliche Verfallstage)
    - Fed-Sitzungen (FOMC Meetings)
    - Feiertags-Effekte
"""
from __future__ import annotations

import pandas as pd
import streamlit as st
from typing import Optional


def compute_streaks_from_df(
    df: pd.DataFrame,
    group_col: str,
    return_col: str = "return",
    label_map: Optional[dict] = None,
    sort_groups: bool = True,
) -> list[dict]:
    """
    Berechnet Streak-Daten aus einem DataFrame.

    Args:
        df: DataFrame mit mindestens group_col und return_col
        group_col: Spalte fuer Gruppierung (z.B. "weekday", "month", "phase")
        return_col: Spalte mit Renditen (default: "return")
        label_map: Dict {group_value: "Anzeigename"} (optional)
        sort_groups: Gruppen sortieren (default: True)

    Returns:
        Liste von Dicts: [{"label": str, "entries": [{"date": ..., "ret": float}]}]
    """
    if df is None or df.empty or group_col not in df.columns:
        return []

    _df = df.dropna(subset=[return_col]).copy()
    groups = sorted(_df[group_col].unique()) if sort_groups else list(_df[group_col].unique())

    result = []
    for g in groups:
        g_df = _df[_df[group_col] == g].sort_index(ascending=False)
        entries = [
            {"date": idx, "ret": float(row[return_col]) * 100 if abs(row[return_col]) < 1 else float(row[return_col])}
            for idx, row in g_df.iterrows()
            if pd.notna(row[return_col])
        ]
        label = label_map[g] if label_map and g in label_map else str(g)
        result.append({"label": label, "entries": entries})

    return result


def compute_streaks_from_list(
    data: list[dict],
    group_key: str,
    return_key: str,
    year_key: str = "year",
    label_map: Optional[dict] = None,
) -> list[dict]:
    """
    Berechnet Streak-Daten aus einer Liste von Dicts (z.B. TOM-Ergebnisse).

    Args:
        data: Liste von Dicts mit group_key, return_key, year_key
        group_key: Key fuer Gruppierung (z.B. "month")
        return_key: Key fuer Rendite (z.B. "total_return")
        year_key: Key fuer Sortierung (z.B. "year")
        label_map: Dict {group_value: "Anzeigename"} (optional)

    Returns:
        Liste von Dicts: [{"label": str, "entries": [{"date": year, "ret": float}]}]
    """
    grouped = {}
    for entry in data:
        g = entry[group_key]
        if g not in grouped:
            grouped[g] = []
        grouped[g].append({"date": entry.get(year_key, ""), "ret": entry[return_key]})

    result = []
    for g in sorted(grouped.keys()):
        entries = sorted(grouped[g], key=lambda x: x["date"], reverse=True)
        label = label_map[g] if label_map and g in label_map else str(g)
        result.append({"label": label, "entries": entries})

    return result


def render_streak_table(
    groups: list[dict],
    col_header: str = "Gruppe",
    n_blocks: int = 10,
    interpretation: str = "",
):
    """
    Rendert eine Streak-Tabelle als HTML in Streamlit.

    Args:
        groups: Liste von {"label": str, "entries": [{"date": ..., "ret": float}]}
        col_header: Header fuer die erste Spalte (z.B. "Wochentag", "Monatswechsel")
        n_blocks: Anzahl der W/L-Bloecke (default: 10)
        interpretation: Optionaler Interpretationstext unter der Tabelle
    """
    if not groups:
        st.info("Keine Daten fuer Streak-Analyse verfuegbar.")
        return

    streak_rows = []
    for group in groups:
        entries = group["entries"]
        if not entries:
            continue

        # Aktuelle Streak zaehlen
        streak_type = "win" if entries[0]["ret"] > 0 else "loss"
        streak_count = 0
        for e in entries:
            if (streak_type == "win" and e["ret"] > 0) or (streak_type == "loss" and e["ret"] <= 0):
                streak_count += 1
            else:
                break

        # W/L Bloecke
        blocks = ""
        for e in entries[:n_blocks]:
            color = "#00d4aa" if e["ret"] > 0 else "#ff4757"
            wl = "W" if e["ret"] > 0 else "L"
            # Tooltip: Datum + Rendite
            if hasattr(e["date"], "strftime"):
                d_str = e["date"].strftime("%d.%m.%y")
            else:
                d_str = str(e["date"])
            ret_fmt = "{:+.2f}%".format(e["ret"])
            blocks += ("<span style='display:inline-block; width:36px; height:28px; "
                       "background:{}; border-radius:5px; margin:2px; "
                       "text-align:center; font-size:11px; line-height:28px; "
                       "font-weight:700; color:#FFFFFF;' "
                       "title='{}: {}'>{}</span>".format(color, d_str, ret_fmt, wl))

        streak_color = "#00d4aa" if streak_type == "win" else "#ff4757"
        streak_text = "{}x {}".format(streak_count, "Gewinn" if streak_type == "win" else "Verlust")

        streak_rows.append(
            "<tr>"
            "<td style='color:#FFFFFF; font-size:13px; padding:6px 12px;'>{}</td>"
            "<td style='color:{}; font-weight:700; font-size:13px; "
            "padding:6px 12px; text-align:center;'>{}</td>"
            "<td style='padding:6px 8px;'>{}</td>"
            "</tr>".format(group["label"], streak_color, streak_text, blocks)
        )

    table_html = (
        "<table style='width:100%; border-collapse:collapse;'>"
        "<tr style='border-bottom:1px solid rgba(255,255,255,0.1);'>"
        "<th style='color:#8899aa; font-size:11px; text-align:left; padding:4px 12px;'>{}</th>"
        "<th style='color:#8899aa; font-size:11px; text-align:center; padding:4px 12px;'>Aktuelle Serie</th>"
        "<th style='color:#8899aa; font-size:11px; text-align:left; padding:4px 8px;'>Letzte {} (neueste links)</th>"
        "</tr>"
        "{}"
        "</table>".format(col_header, n_blocks, "".join(streak_rows))
    )
    st.markdown(table_html, unsafe_allow_html=True)

    if interpretation:
        st.markdown(
            "<p style='color:#FFFFFF; font-size:12px; margin-top:12px; line-height:1.6;'>"
            "<b>Interpretation:</b> {}</p>".format(interpretation),
            unsafe_allow_html=True)

File: landing/js/streak-analysis.js
/**
 * SeasonAlpha — Streak Analysis (Port von shared/streak_analysis.py)
 * ===================================================================
 * W/L-Serien Berechnung + HTML-Tabelle.
 * Wird von: Monatswechsel, Wochentage, Mondphasen genutzt.
 */

var SA = window.SA || {};

SA.streaks = {

  /**
   * Berechnet Streaks aus TOM/Event-Ergebnislisten.
   * (Port von compute_streaks_from_list)
   *
   * @param {Array} data - [{year, month, total_return}, ...]
   * @param {string} groupKey - Feld zum Gruppieren (z.B. "month")
   * @param {string} returnKey - Feld mit Rendite (z.B. "total_return")
   * @param {string} yearKey - Feld zum Sortieren (z.B. "year")
   * @param {Object} labelMap - {1: "Jan → Feb", ...} (optional)
   * @returns {Array} [{label, entries: [{date, ret}]}]
   */
  computeStreaksFromList: function(data, groupKey, returnKey, yearKey, labelMap) {
    yearKey = yearKey || 'year';
    var groups = {};
    for (var i = 0; i < data.length; i++) {
      var d = data[i];
      var g = d[groupKey];
      if (!groups[g]) groups[g] = [];
      groups[g].push({ date: d[yearKey], ret: d[returnKey] });
    }

    var result = [];
    var sortedKeys = Object.keys(groups).sort(function(a, b) { return a - b; });
    for (var ki = 0; ki < sortedKeys.length; ki++) {
      var key = sortedKeys[ki];
      var entries = groups[key].slice();
      // Sortieren: neueste zuerst
      entries.sort(function(a, b) { return b.date - a.date; });
      var label = labelMap ? (labelMap[key] || 'x' + key) : 'x' + key;
      result.push({ label: label, entries: entries });
    }
    return result;
  },

  /**
   * Aktuelle Streak zaehlen (vom neuesten Eintrag aus).
   * @param {Array} entries - [{date, ret}] (neueste zuerst)
   * @returns {Object} {type: "win"|"loss", count: N}
   */
  currentStreak: function(entries) {
    if (!entries || entries.length === 0) return { type: 'none', count: 0 };
    var type = entries[0].ret > 0 ? 'win' : 'loss';
    var count = 1;
    for (var i = 1; i < entries.length; i++) {
      var isWin = entries[i].ret > 0;
      if ((type === 'win' && isWin) || (type === 'loss' && !isWin)) {
        count++;
      } else {
        break;
      }
    }
    return { type: type, count: count };
  },

  /**
   * Rendert eine Streak-Tabelle als HTML.
   * (Port von render_streak_table)
   *
   * @param {Array} groups - [{label, entries}] (von computeStreaksFromList)
   * @param {string} colHeader - Spaltenheader (default "Gruppe")
   * @param {number} nBlocks - Anzahl W/L-Bloecke (default 10)
   * @returns {string} HTML-String
   */
  renderStreakTable: function(groups, colHeader, nBlocks) {
    var _en = !!(SA.i18n && SA.i18n.isEN && SA.i18n.isEN());
    colHeader = colHeader || (_en ? SA.i18n.t('str.col_group') : 'Gruppe');
    nBlocks = nBlocks || 10;
    var C = SA.COLORS || { green: '#00d4aa', red: '#ff4757' };

    var html = '<table data-no-sort="1" style="width:100%;border-collapse:collapse;font-size:.8125rem">';
    html += '<tr><th style="text-align:left;padding:.5rem;color:var(--muted,#a89878);border-bottom:1px solid rgba(255,255,255,.06)">' + colHeader + '</th>';
    html += '<th style="text-align:left;padding:.5rem;color:var(--muted,#a89878);border-bottom:1px solid rgba(255,255,255,.06)">' + (_en ? SA.i18n.t('str.current_streak') : 'Aktuelle Serie') + '</th>';
    html += '<th style="text-align:left;padding:.5rem;color:var(--muted,#a89878);border-bottom:1px solid rgba(255,255,255,.06)">' + (_en ? SA.i18n.t('str.last_n').replace('{n}', nBlocks) : 'Letzte ' + nBlocks) + '</th></tr>';

    for (var gi = 0; gi < groups.length; gi++) {
      var g = groups[gi];
      var streak = SA.streaks.currentStreak(g.entries);
      var streakColor = streak.type === 'win' ? C.green : C.red;
      var streakText = streak.count + 'x ' + (streak.type === 'win'
        ? (_en ? SA.i18n.t('str.win') : 'Gewinn')
        : (_en ? SA.i18n.t('str.loss') : 'Verlust'));

      // W/L Bloecke (chronologisch: aelteste links, neueste rechts)
      var blocks = g.entries.slice(0, nBlocks).reverse();
      var blocksHtml = '';
      for (var bi = 0; bi < blocks.length; bi++) {
        var b = blocks[bi];
        var isWin = b.ret > 0;
        var bgColor = isWin ? C.green : C.red;
        var label = isWin ? 'W' : 'L';
        var retFmt = (b.ret >= 0 ? '+' : '') + b.ret.toFixed(2) + '%';
        blocksHtml += '<span title="' + b.date + ': ' + retFmt + '" style="display:inline-block;width:32px;height:26px;line-height:26px;text-align:center;background:' + bgColor + ';color:#fff;font-size:10px;font-weight:700;border-radius:4px;margin:1px">' + label + '</span>';
      }

      html += '<tr style="border-bottom:1px solid rgba(255,255,255,.03)">';
      html += '<td style="padding:.4rem .5rem;color:var(--dim,#e8e0d0)">' + g.label + '</td>';
      html += '<td style="padding:.4rem .5rem;color:' + streakColor + ';font-weight:600">' + streakText + '</td>';
      html += '<td style="padding:.4rem .5rem">' + blocksHtml + '</td>';
      html += '</tr>';
    }
    html += '</table>';
    return html;
  }
};

window.SA = SA;
